### PR TITLE
Check IP family for exgws

### LINKED ISSUE #1835: Check IP family for exgws (dualstack)
If someone annotates a namespace or pod with IPs of a certain family (v4 or v6), we need to ensure that when we add the routes for the pods that gateway that we only use the correct IP family addr of the pod. We may also need to create a reject ACL for traffic that comes from the pod in an exgw namespace that is missing a gateway in a particular ip family.

We also need to add e2e tests to ensure this works.

References:
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/egressgw.go#L108
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/egressgw.go#L200
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/egressgw.go#L263

### PR DESCRIPTION
**Pull request by rcarrillocruz**
We need to ensure we only add the proper family routes when exgw
annotations are used.

Fixes #1835

### CODE REVIEW

File: go-controller/pkg/ovn/egressgw.go [hunk 1]
```diff
@@ -106,6 +106,9 @@ func (oc *Controller) addGWRoutesForNamespace(namespace string, gws []net.IP, ns
 		gr := "GR_" + pod.Spec.NodeName
 		for _, gw := range gws {
 			for _, podIP := range pod.Status.PodIPs {
+				if (utilnet.IsIPv4(gw) && utilnet.IsIPv6(net.ParseIP(podIP.IP))) || (utilnet.IsIPv6(gw) && utilnet.IsIPv4(net.ParseIP(podIP.IP))) {
```
Comment (aojea) [inline go-controller/pkg/ovn/egressgw.go:109]: I think that you can simplify it

`if utilnet.IsIPv6(gw) != utilnet.IsIPv6String(PodIP.IP) `

Comment (rcarrillocruz) [reply go-controller/pkg/ovn/egressgw.go:109]: I prefer more readable even if it's more verbose, but yeah, I come from Python, I don't have a strong opinion, will change.

Comment (danwinship) [reply go-controller/pkg/ovn/egressgw.go:109]: Yeah, what he said. Likewise below. (Also, huh, I hadn't seen that they added `IsIPv4` now too... there used to only be `IsIPv6`.)

**Comment by rcarrillocruz** [comment]
need to check with @nerdalert with how he'd like to test this

**Comment by nerdalert** [comment]
> need to check with @nerdalert with how he'd like to test this

@rcarrillocruz the PR #1881 tests whether the address family matches between the pod and external gateway. PR lgtm, ty!

**Comment by trozet** [comment]
I didn't realize @nerdalert already added a semi-fix for this before: https://github.com/ovn-org/ovn-kubernetes/pull/1855

**Comment by trozet** [comment]
@nerdalert fix #1855 fixes the path where the pod is added and the multiple gateways already exist for the namespace. This fixes the path where the pods already exist in a namespace that does not have multiple gateways set, but then later the namespace is updated with the multiple gateway annotations. We need test coverage for both of these scenarios @nerdalert can you include that in your #1881 PR please?

**Review by trozet — CHANGES_REQUESTED** [review]
we are missing checking for address family during SNAT creation on the GR, when you use --disable-snat-multiple-gws in KIND:

I0113 17:10:45.081437      52 event.go:282] Event(v1.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"coredns-f9fd979d6-2lpwq", UID:"445f6920-a1df-4f53-9d9e-9e8e9d7a3aeb", APIVersion:"v1", ResourceVersion:"1010", FieldPath:""}): type: 'Warning' reason: 'ErrorAddingLogicalPort' failed to create SNAT rule for pod on gateway router GR_ovn-worker, stdout: "", stderr: "ovn-nbctl: fd00:10:244:3::3/128: should be an IPv4 address or network.\n", error: OVN command '/usr/bin/ovn-nbctl --timeout=15 --if-exists lr-nat-del GR_ovn-worker snat fd00:10:244:3::3/128 -- lr-nat-add GR_ovn-worker snat 172.18.0.4 fd00:10:244:3::3/128' failed: exit status 1

https://github.com/ovn-org/ovn-kubernetes/blob/c85526c1bcf8c4fb141acc33218cb13319761015/go-controller/pkg/ovn/egressgw.go#L297

https://github.com/ovn-org/ovn-kubernetes/blob/c85526c1bcf8c4fb141acc33218cb13319761015/go-controller/pkg/ovn/egressgw.go#L310

**Comment by trozet** [comment]
Looks good, I see both paths working. Updating namespace to have multiple gateways:
```
I0113 19:41:16.073315      51 namespace.go:213] Updating namespace: exgw
I0113 19:41:16.074612      51 ovs.go:166] exec(269): /usr/bin/ovn-nbctl --timeout=15 -- --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 10.244.2.3/32 172.18.0.233
I0113 19:41:16.079755      51 ovs.go:169] exec(269): stdout: ""
I0113 19:41:16.079789      51 ovs.go:170] exec(269): stderr: ""
I0113 19:41:16.079884      51 ovs.go:166] exec(270): /usr/bin/ovn-nbctl --timeout=15 -- --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 fd00:10:244:3::3/128 fc00:f853:ccd:e793::233

```

Adding new pod in gw namespace:
```
I0113 19:43:54.846591      51 pods.go:254] Creating logical port for exgw_pod2 on switch ovn-worker2
I0113 19:43:54.855146      51 pods.go:393] Annotation values: ip=[10.244.2.4/24 fd00:10:244:3::4/64] ; mac=0a:58:0a:f4:02:04 ; gw=[10.244.2.1 fd00:10:244:3::1]
Annotation=map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.244.2.4/24","fd00:10:244:3::4/64"],"mac_address":"0a:58:0a:f4:02:04","gateway_ips":["10.244.2.1","fd00:10:244:3::1"]}}]
I0113 19:43:54.855710      51 kube.go:63] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.244.2.4/24","fd00:10:244:3::4/64"],"mac_address":"0a:58:0a:f4:02:04","gateway_ips":["10.244.2.1","fd00:10:244:3::1"]}}] on pod exgw/pod2
I0113 19:43:54.864506      51 port_cache.go:55] port-cache(exgw_pod2): added port &{name:exgw_pod2 uuid:6c0eec22-c0fd-4f80-8f31-02608bb0bd08 logicalSwitch:ovn-worker2 ips:[0xc00091aae0 0xc00091ab10] mac:[10 88 10 244 2 4] expires:{wall:0 ext:0 loc:<nil>}}
I0113 19:43:54.864788      51 address_set.go:418] (a65625d0-076b-4763-836a-63536e42d234/exgw_v4/a11491463288749249233) adding IP 10.244.2.4 to address set
I0113 19:43:54.864869      51 ovs.go:166] exec(271): /usr/bin/ovn-nbctl --timeout=15 add address_set a65625d0-076b-4763-836a-63536e42d234 addresses "10.244.2.4"
I0113 19:43:54.872524      51 ovs.go:169] exec(271): stdout: ""
I0113 19:43:54.872565      51 ovs.go:170] exec(271): stderr: ""
I0113 19:43:54.872590      51 address_set.go:418] (0ff3c317-fd05-4599-a43b-866360c6a46b/exgw_v6/a11491461089725992811) adding IP fd00:10:244:3::4 to address set
I0113 19:43:54.872667      51 ovs.go:166] exec(272): /usr/bin/ovn-nbctl --timeout=15 add address_set 0ff3c317-fd05-4599-a43b-866360c6a46b addresses "fd00:10:244:3::4"
I0113 19:43:54.879577      51 ovs.go:169] exec(272): stdout: ""
I0113 19:43:54.879617      51 ovs.go:170] exec(272): stderr: ""
I0113 19:43:54.879716      51 ovs.go:166] exec(273): /usr/bin/ovn-nbctl --timeout=15 --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 10.244.2.4/32 172.18.0.233
I0113 19:43:54.884099      51 ovs.go:169] exec(273): stdout: ""
I0113 19:43:54.884130      51 ovs.go:170] exec(273): stderr: ""
I0113 19:43:54.884209      51 ovs.go:166] exec(274): /usr/bin/ovn-nbctl --timeout=15 --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 fd00:10:244:3::4/128 fc00:f853:ccd:e793::233
I0113 19:43:54.888199      51 ovs.go:169] exec(274): stdout: ""
I0113 19:43:54.888232      51 ovs.go:170] exec(274): stderr: ""
I0113 19:43:54.888927      51 pods.go:244] [exgw/pod2] addLogicalPort took 43.294067ms

```

**Comment by trozet** [comment]
and updating a host networked pod with annotations also works:
```
I0113 19:47:55.207936      51 egressgw.go:28] External gateway pod: pod1, detected for namespace(s) exgw
I0113 19:47:55.214380      51 ovs.go:166] exec(275): /usr/bin/ovn-nbctl --timeout=15 -- --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 10.244.2.3/32 172.18.0.99
I0113 19:47:55.242635      51 ovs.go:169] exec(275): stdout: ""
I0113 19:47:55.242736      51 ovs.go:170] exec(275): stderr: ""
I0113 19:47:55.242957      51 ovs.go:166] exec(276): /usr/bin/ovn-nbctl --timeout=15 -- --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 fd00:10:244:3::3/128 fc00:f853:ccd:e793::266
I0113 19:47:55.259552      51 ovs.go:169] exec(276): stdout: ""
I0113 19:47:55.259627      51 ovs.go:170] exec(276): stderr: ""
I0113 19:47:55.259782      51 ovs.go:166] exec(277): /usr/bin/ovn-nbctl --timeout=15 -- --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 10.244.2.4/32 172.18.0.99
I0113 19:47:55.272254      51 ovs.go:169] exec(277): stdout: ""
I0113 19:47:55.272299      51 ovs.go:170] exec(277): stderr: ""
I0113 19:47:55.272399      51 ovs.go:166] exec(278): /usr/bin/ovn-nbctl --timeout=15 -- --may-exist --policy=src-ip --ecmp-symmetric-reply lr-route-add GR_ovn-worker2 fd00:10:244:3::4/128 fc00:f853:ccd:e793::266
I0113 19:47:55.285542      51 ovs.go:169] exec(278): stdout: ""
I0113 19:47:55.285583      51 ovs.go:170] exec(278): stderr: ""

```

**Pull request merged**